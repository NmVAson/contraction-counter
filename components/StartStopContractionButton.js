import React, {Component} from 'react';
import {StyleSheet, View} from 'react-native';
import Button from "react-native-bootstrap-buttons";

export default class StartStopContractionButton extends Component {
  constructor(props) {
    super(props);
    this.state = { 
      title: 'click to start contraction',
      isContracting: false
    };
  }

  toggleContraction() {
    let isStartingANewContraction = !this.state.isContracting;
    let newTitle;

    if(isStartingANewContraction) {
      this.props.startContraction();
      newTitle = 'click to end contraction';
    } else {
      this.props.endContraction();
      newTitle = 'click to start contraction';
    }

    this.setState((state, props) => ({
      title: newTitle,
      isContracting: !state.isContracting
    }));
  }

  render() {
    return (<Button 
        label={this.state.title}
        labelStyle={styles.buttonLabel}
        onPress={this.toggleContraction.bind(this)}/>);
  }
}

const styles = StyleSheet.create({
  buttonLabel: { 
    fontSize: 20, 
    padding: 20,
    textAlign: 'center'
  }
});
