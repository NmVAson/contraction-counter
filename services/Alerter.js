import React, {Component} from 'react';
import {Alert} from 'react-native';

const oneHourInMs = 3600000;
const delta = 0.15;

export default class Alerter {
    static trackContraction(frequency, duration) {
        let isLessThanFiveMinutesApart = frequency <= 5 + delta;
        let isAboutAMinute = Math.abs(duration - 1) <= delta;
        let isTimeToCallDoc = isLessThanFiveMinutesApart && isAboutAMinute;

        if(isTimeToCallDoc) {
            Alert.alert('Time to call your doctor!', 'You\'ll be heading to the hospital in an hour!');
            setTimeout(() => Alert.alert('Time to head to the hospital!', 'Don\'t forget your hospital bag :)'), oneHourInMs);
        }
    }
}