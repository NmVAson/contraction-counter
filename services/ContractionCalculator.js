import React, {Component} from 'react';
import {Alert} from 'react-native';
import moment from 'moment';

const oneHourInMs = 3600000;

export default class ContractionCalculator {
    static getDurationInMinutes(contraction) {
        let {start, end} = contraction;
        let diff = contraction.end.diff(contraction.start);
        let duration = moment.duration(diff);

        return duration.asMinutes();
    }

    static getDurationToDisplay(contraction) {
        let {start, end} = contraction;
        let diff = moment(contraction.end).diff(moment(contraction.start));
        let duration = moment.duration(diff).asMilliseconds();

        return moment.utc(duration).format("m[m] s[s]");
    }

    static getFrequencyInMinutes(previousContraction, currentContraction) {
        let frequencyDiff = moment(currentContraction.start).diff(moment(previousContraction.start));
        let frequency = moment.duration(frequencyDiff);

        return frequency.asMinutes();
    }

    static getFrequencyToDisplay(previousContraction, currentContraction) {
        let frequencyDiff = moment(currentContraction.start).diff(moment(previousContraction.start));
        let frequency = moment.duration(frequencyDiff).asMilliseconds();

        return moment.utc(frequency).format("H[h] m[m] s[s]")
    }
}