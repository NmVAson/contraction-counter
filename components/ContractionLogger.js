import React, {Component} from 'react';
import {View, Text, AsyncStorage} from 'react-native';
import moment from 'moment';
import shortid from 'shortid';

import Alerter from '../services/Alerter';
import ContractionCalculator from '../services/ContractionCalculator';
import StartStopContractionButton from './StartStopContractionButton';
import ContractionList from './ContractionList';
import { Context } from '../context';


export default function ContractionLogger({ styles }) {
    const {contractions, addContraction} = React.useContext(Context);
  
    onContractionStart = () => {
      addContraction({start: moment(), end: null});
    }
  
    onContractionEnd = () => {
      let currentContraction = contractions.pop();
      currentContraction.end = moment();

      if(contractions.length >= 1) {
        let previousContraction = contractions.slice(-1).pop();
        let duration = ContractionCalculator.getDurationInMinutes(currentContraction);
        let frequency = ContractionCalculator.getFrequencyInMinutes(previousContraction, currentContraction);

        Alerter.trackContraction(frequency, duration);
      }

      addContraction(currentContraction);
    }
  
    return (<View style={[styles.container, styles.h80]}>
        <StartStopContractionButton
            startContraction={this.onContractionStart.bind(this)}
            endContraction={this.onContractionEnd.bind(this)}/>
        <ContractionList/>
    </View>);
}
