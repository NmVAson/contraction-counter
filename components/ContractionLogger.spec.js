import React from 'react';
import {View} from 'react-native';
import moment from 'moment';

import StartStopContractionButton from './StartStopContractionButton';
import ContractionLogger from './ContractionLogger';
import Alerter from '../services/Alerter';
import ContractionCalculator from '../services/ContractionCalculator';

import TestRenderer from 'react-test-renderer';

const fakeStyles = {};
const addSpy = jest.fn();
const fakeContext = {
  contractions: [],
  addContraction: addSpy.mockImplementation((contraction) => fakeContext.contractions.push(contraction))
};
const now = moment();
let component;
let testRenderer;

beforeEach(() => {
  jest.spyOn(Date, 'now').mockImplementation(() => now);
  jest.spyOn(React, 'useContext').mockImplementation(() => fakeContext);

  testRenderer = TestRenderer.create(<ContractionLogger styles={fakeStyles}/>);
  component = testRenderer.root.findByType(ContractionLogger);
});

it('matches previous snapshot', () => {
  let tree = TestRenderer
    .create(<ContractionLogger styles={fakeStyles}/>)
    .toJSON();
  
  expect(tree).toMatchSnapshot();
});

it('supplies a new contraction function to button', () => {
  let button = component.findByType(StartStopContractionButton);
  let expectedContraction = {
    start: now,
    end: null
  };

  button.props.startContraction();

  expect(addSpy).toHaveBeenCalled();
  expect(addSpy).toHaveBeenCalledWith(expectedContraction);
});

it('supplies an end contraction function to button', () => {
  let button = component.findByType(StartStopContractionButton);
  let expectedContraction = {
    start: now,
    end: now
  };
  button.props.startContraction();
  addSpy.mockClear();

  button.props.endContraction();

  expect(addSpy).toHaveBeenCalled();
  expect(addSpy).toHaveBeenCalledWith(expectedContraction);
});

it('tracks contractions for alerts', () => {
  let expectedDuration = 1;
  let expectedFrequency = 2;
  let button = component.findByType(StartStopContractionButton);
  let alerterSpy = jest.spyOn(Alerter, 'trackContraction');
  jest.spyOn(ContractionCalculator, 'getDurationInMinutes').mockImplementation(() => expectedDuration);
  jest.spyOn(ContractionCalculator, 'getFrequencyInMinutes').mockImplementation(() => expectedFrequency);

  button.props.startContraction();
  button.props.endContraction();
  button.props.startContraction();
  button.props.endContraction();

  expect(alerterSpy).toHaveBeenCalled();
  expect(alerterSpy).toHaveBeenCalledWith(expectedFrequency, expectedDuration);
});
