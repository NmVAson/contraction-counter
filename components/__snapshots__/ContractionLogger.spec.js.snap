// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches previous snapshot 1`] = `
<View
  style={
    Array [
      undefined,
      undefined,
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#0984e3",
        "borderRadius": 5,
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "padding": 10,
          },
          Object {
            "color": "#ffffff",
          },
          Object {
            "fontSize": 20,
            "padding": 20,
            "textAlign": "center",
          },
        ]
      }
    >
      click to start contraction
    </Text>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "flex": 1,
        "flexDirection": "column",
        "paddingTop": 30,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#f1f8ff",
          "flex": 0,
          "flexDirection": "row",
          "height": 70,
          "padding": 20,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "column",
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 20,
            }
          }
        >
          Frequency
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "column",
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 20,
            }
          }
        >
          Duration
        </Text>
      </View>
    </View>
    <RCTScrollView
      onContentSizeChange={[Function]}
    >
      <View />
    </RCTScrollView>
  </View>
</View>
`;
